### DISCUSSION TITLE
Workflow for running unit tests locally

### DISCUSSION THREAD

**Discussion by BrownianNotion** [musescore/MuseScore#22731] [Q&A] [answered]
Hi new contributor here,
I have spent some considerable time looking around the repo and wiki/discussions, but am still unsure what the standard way of running unit tests locally is. 

I am building in the command line with cmake - I have got the unit tests running on my machine with the following:
```
cmake -P build.cmake test
cd builds/Linux-Qt6.2.4-gcc_64-Make-Testing
ctest --output-on-failure
```
I am quite sure I have already followed all the instructions on the musescore wiki, especially the steps listed [here](https://github.com/musescore/MuseScore/wiki/Compile-on-the-command-line).

**Two main questions:**
1. Is the above the best way to run unit tests locally? I've looked into `buildscripts/ci/linux/local/workflow.sh` which is meant to replicate the ci worfklow setup locally but I'm not sure if this is up to date, and I cannot get it to work (various different errors like broken paths).
2. What is the best way to run the vtests locally? Have fiddled around with `vtest/vtest-compare-pngs.sh` which has broken paths (mainly relating to `build.debug` which doesn't seem to exist), but have not gotten it to work yet. Fixing the paths to the binary helped, but I cannot seem to generate the reference pngs properly. Before I spend any further time trying to figure this out, I thought it'd be better to ask here.

Any help is highly appreciated!

**Comment by shoogle** [comment, marked as answer]
Welcome!

1. See [Building the unit tests locally](https://github.com/musescore/MuseScore/wiki/Fix-the-CI-checks#building-the-unit-tests-locally). Building on the command line is recommended for your first time compiling, but once you've done that successfully we recommend compiling in Qt Creator.
2. I'm not sure if anyone builds the vtests locally. It's probably best to ask in the [Discord Server](https://github.com/musescore/MuseScore/wiki/Contact#MuseScores-Discord-server)'s _#development_ channel.

**Reply by BrownianNotion** [reply]
Thanks, will look into qtcreator! Just while we're here, is there much difference between what you suggested, vs either of the two options below? 
```
cmake -P build.cmake test
```

```
cmake -P build.cmake -DCMAKE_BUILD_TYPE=Testing 
```

**Reply by shoogle** [reply]
Neither of those arguments are valid.

- The `build.cmake` script doesn't have an option called `test` (although arguably we should add this option).

- The CMake variable `CMAKE_BUILD_TYPE` can only take the values mentioned [here](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html), and `Testing` is not one of the allowed values.

The unit tests are compiled by default, so you don't need to enable any special options to compile them. You simply compile the program the normal way and then run `ctest` in the build directory, or via Qt Creator.

**Reply by BrownianNotion** [reply]
Thanks :) my mistake
